### ISSUE TITLE
Autocomplete for DI Containers

### ISSUE THREAD

**Issue by brad-jones** [php-integrator/atom-autocompletion#83]
These days it is fairly common to ask for an instance of a class from a container like this:

```php
$container->get(Foo::class)->bar...

// OR

$container[Foo::class]->bar...
```

It would be nice if these cases were supported.

Perhaps you can just check if `$container` implements [ContainerInterface](https://github.com/container-interop/container-interop/blob/master/src/Interop/Container/ContainerInterface.php) and if so make the assumption that what ever is returned by `get()` or the array access style will be an instance of `Foo::class`.

**Comment by Gert-dev** [comment]
This is a valid suggestion and indeed currently a problem. The issue here is that the type of the returned value depends on method parameters.

Even though `ContainerInterface` appears to be a somewhat PSR-like library, I'm not fond of introducing library- or framework-specific code into the core. If I allow this now, I'd expect issues for various frameworks (Symfony, Laravel, ...) or custom-built service locators to show up sooner or later and I'd be in a rather weak position to reject them solely on the basis that they are too framework-specific.

It is currently already possible to solve this in another way, but it's rather cumbersome if you need services often:

```php
/** @var Foo $fooService */
$fooService = $container->get(Foo::class);
$fooService-> // Should autocompete for Foo.
```

You'd need to repeat this everytime you fetch the service (hence why it's cumbersome). However, there is a way to solve this problem using a more generic format that also applies to things other than service locators which you may have heard of: PhpStorm meta files [1]. In these files you can apply these hints over your entire code base and you can generically inform the type analyzer that `if string value S is passed to method M of class C, it returns an instance of class I`, which can be used for service locators, but also for other classes and requires no specific integration.

Seeing as not every framework provides a way to generate these files automatically based on their service locator, it may still be interesting to allow extensions in the future that would plug into the core to support libraries such as `ContainerInterface`, the Symfony container, etc. but I'd need to establish an extension path for those kind of libraries first. It would also solve the problem of introducing them into the core as they would likely be separate libraries that are independent from it.

In short, your suggestion is certainly relevant and I plan on at least supporting PhpStorm meta files at some point in the future (and hopefully extensions sometime after that)!

[1] https://github.com/php-integrator/core/issues/10

**Comment by brad-jones** [comment]
@Gert-dev I understand where you are coming from.
I think PhpStorm meta files provide the most flexible solution.

**Comment by Gert-dev** [comment]
@brad-jones Alright, I'll close this ticket since it's already taken up in the core ticket.
